### PR TITLE
Abstract based plugin manager

### PR DESCRIPTION
**Pull request by Lanpingner**
**Description**
Very basic plugin manager to load/unload plugins:

**Todo**

- [x] Abstract based plugin manager (for later API use)
- [x]  Discord moved to as a plugin
- [x] Railroad activity that can notify plugins
- [x] Config read for plugins
- [x] Threads for plugin health check / plugin based

**Example output**
`
[PluginManager][io.github.railroad.Plugins.Discord] Change state from: NOTLOADED to: STARTING_INIT
[PluginManager][io.github.railroad.Plugins.Discord] Errors: [java.lang.Error: Failed to initialize Discord IPC channel]
`

### CODE REVIEW

File: src/main/java/io/github/railroad/PluginManager/Plugin.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private PluginStates state = PluginStates.NOTLOADED;
+    private PluginHealthChecker healthChecker;
+
+    public abstract PluginPhaseResult InitPlugin();
+
+    public abstract PluginPhaseResult LoadPlugin();
+
+    public abstract PluginPhaseResult UnloadPlugin();
+
+    public abstract PluginPhaseResult RaildraodActivityChange(RailroadActivities.RailroadActivityTypes railroadActivityTypes);
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/Plugin.java]: Misspelling

File: src/main/java/io/github/railroad/PluginManager/Plugin.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    public abstract PluginPhaseResult ReloadPlugin();
+
+    public void UpdateStatus(PluginStates state) {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/Plugin.java]: Should be `camelCase` for methods.

File: src/main/java/io/github/railroad/PluginManager/Plugin.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public PluginStates getState() {
+        return this.state;
+    }
+
+    public PluginPhaseResult getNewPhase() {
+        return new PluginPhaseResult();
+    }
+
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/Plugin.java:38]:
```suggestion
```

File: src/main/java/io/github/railroad/PluginManager/PluginHealthChecker.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+public class PluginHealthChecker extends Thread {
+    private Plugin plugin;
+
+    public PluginHealthChecker(Plugin plugin) {
+        this.plugin = plugin;
+        this.setName("HC_" + plugin.getClass().getName());
+    }
+
+    public void run() {
+
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/PluginHealthChecker.java]:
```suggestion
```

File: src/main/java/io/github/railroad/PluginManager/PluginHealthChecker.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+                }
+                sleep(10000);
+            } catch (InterruptedException e) {
+                PluginPhaseResult result = new PluginPhaseResult();
+                result.AddError(new Error(e.getMessage()));
+                plugin.getPluginManager().showError(plugin, result, "Health check loop error");
+            }
+        }
+    }
+
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/PluginHealthChecker.java]:
```suggestion
```

File: src/main/java/io/github/railroad/PluginManager/PluginManager.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                    showError(plugin, loadphaseResult, "LoadPlugin");
+                }
+            } else {
+                showError(plugin, initphaseResult, "InitPlugin");
+            }
+
+        }
+    }
+
+    private void PrepareluginsFromConfig() {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/PluginManager.java]: Misspelling

File: src/main/java/io/github/railroad/PluginManager/PluginManager.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import io.github.railroad.discord.activity.RailroadActivities;
+import io.github.railroad.utility.ConfigHandler;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class PluginManager extends Thread {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/PluginManager.java:13]: Same `camelCase` for methods thing here.

File: src/main/java/io/github/railroad/PluginManager/PluginStates.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package io.github.railroad.PluginManager;
+
+public enum PluginStates {
+    NOTLOADED, STARTING_INIT, ERROR_INIT, FINSIHED_INIT, LOADING, LOADED, ACTIVITY_UPDATE_START, ACTIVITY_UPDATE_FINSIHED, ACTIVITY_UPDATE_ERROR
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/PluginManager/PluginStates.java]: `NOT_LOADED`

File: src/main/java/io/github/railroad/Plugins/Discord.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        PluginPhaseResult phaseResult = this.getNewPhase();
+        try {
+            var discord = new DiscordCore("853387211897700394");
+
+            Runtime.getRuntime().addShutdownHook(new Thread(discord::close));
+            this.DISCORD = discord;
+
+        } catch (Exception e) {
+            this.UpdateStatus(PluginStates.ERROR_INIT);
+            phaseResult.AddError(new Error(e.getMessage()));
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/Plugins/Discord.java]:
```suggestion
        } catch (Exception exception) {
            this.UpdateStatus(PluginStates.ERROR_INIT);
            phaseResult.AddError(new Error(exception.getMessage()));
```

File: src/main/java/io/github/railroad/Railroad.java [hunk 10]
```diff
@@ -29,9 +29,10 @@ public class Railroad extends Application {
     public static final OkHttpClient HTTP_CLIENT = new OkHttpClient();
     public static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
     public static final ProjectManager PROJECT_MANAGER = new ProjectManager();
+    public static final PluginManager PLUGIN_MANAGER = new PluginManager();
 
     private static boolean DEBUG = false;
-    private static DiscordCore DISCORD;
+    ;
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/Railroad.java]:
```suggestion
```

File: src/main/java/io/github/railroad/utility/ConfigHandler.java [hunk 11]
```diff
@@ -12,6 +12,24 @@
 public final class ConfigHandler {
     private ConfigHandler() {}
 
+    public static void CheckAndCreateDefaultJsonObjects(JsonObject object) {
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/railroad/utility/ConfigHandler.java]: `camelCase` for methods.

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/railroad/PluginManager/Plugin.java (hunk 1 above)
Comment (Lanpingner) [reply src/main/java/io/github/railroad/PluginManager/Plugin.java]: Legit

**Review by DaRealTurtyWurty — APPROVED** [review]

**Pull request merged**